### ISSUE TITLE
Erroneous warning in autofunction

### ISSUE THREAD

**Issue by pganssle** [sphinx-doc/sphinx#5431]
### Problem
The change in #5345 seems to have introduced an issue where using the `..autofunction` directive on a callable object that is not a `function` warns about an ignored `TypeError`.

On my [MWE repo](https://github.com/pganssle-bug-mwes/sphinx-autofunc-bug), I get:

```
$ make -C docs html
make: Entering directory '/path/to/sphinx-bug/docs'
Running Sphinx v1.8.0
[... snipped ...]

WARNING: Invalid type annotation found on <autobugdemo.__get_func.<locals>.MyFunction object at 0x7fb3638951d0>. Ingored: TypeError('<autobugdemo.__get_func.<locals>.MyFunction object at 0x7fb3638951d0> is not a module, class, method, or function.')

[.. snipped ..]

The HTML pages are in _build/html.
```

Where `autobugdemo.func` is defined as:

```python
def __get_func():
    class MyFunction:
        def __call__(self):
            print("Hello, world")
    return MyFunction()

func = __get_func()
```

#### Procedure to reproduce the problem
Using my [mwe repo](https://github.com/pganssle-bug-mwes/sphinx-autofunc-bug):
```
$ git clone https://github.com/pganssle-bug-mwes/sphinx-autofunc-bug.git
$ cd sphinx-autofunc-bug
$ make -C docs html
```

#### Error logs / results
The full log (unsnipped) is found below, under the `details` tag:

<details>

```
$ make -C docs html
make: Entering directory '/path/to/sphinx-bug/docs'
Running Sphinx v1.8.0
making output directory...
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 1 source files that are out of date
updating environment: 1 added, 0 changed, 0 removed
reading sources... [100%] index                                                                                                             
WARNING: Invalid type annotation found on <autobugdemo.__get_func.<locals>.MyFunction object at 0x7fb3638951d0>. Ingored: TypeError('<autobugdemo.__get_func.<locals>.MyFunction object at 0x7fb3638951d0> is not a module, class, method, or function.')
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [100%] index                                                                                                              
generating indices... genindex
writing additional pages... search
copying static files... done
copying extra files... done
dumping search index in English (code: en) ... done
dumping object inventory... done
build succeeded, 1 warning.

The HTML pages are in _build/html.
make: Leaving directory '/path/to/sphinx-bug/docs'
```

</details>

#### Expected results
No warning should be raised.

Also, this was somewhat annoying to debug, so ideally if a warning *is* raised, it should come with at least some sort of rudimentary traceback.

### Reproducible project / your project
- https://github.com/pganssle-bug-mwes/sphinx-autofunc-bug

### Environment info
- OS: Linux
- Python version: 3.7
- Sphinx version: 1.8.0

**Comment by tk0miya** [comment]
Note: This is happened only with python 3.6/3.7 and Sphinx-1.8.0.

I confirmed this with following Dockerfile:
```
FROM python:3.7

RUN pip3 install Sphinx
RUN git clone https://github.com/pganssle-bug-mwes/sphinx-autofunc-bug
WORKDIR /sphinx-autofunc-bug
RUN make -C docs html
```

**Comment by tk0miya** [comment]
It seems the example gives non-function object to an argument of `autofunction` directive...

**Comment by pganssle** [comment]
@tk0miya Correct. I am able to reliably reproduce this when passing callable objects to `autofunction`. The fact that they are not `function` objects is, essentially, an implementation detail, but it's not possible to create function subclasses - see [PEP 579](https://www.python.org/dev/peps/pep-0579/) for more information on the various problems with this.

Tangential to this - the way `autofunction` assumes that a given function is always a `function` object (and thus could not have any members or methods) is, IMO, less than ideal. Consider that for my `variants` library, I needed to [write a sphinx extension](https://variants.readthedocs.io/en/latest/sphinx.html) that essentially combines aspects of `autofunction` and `autoclass`. I imagine this also makes it difficult to autodoc an `lru_cache`-ed function, which has methods *on the function object* relating to clearing the cache and such.

**Comment by tk0miya** [comment]
I don't know what is best way for marking up callable. But I agree that the warning is not good for us. So I just post #5447 to prevent it.

**Comment by tk0miya** [comment]
Fixed by #5447.
Thank you for reporting.
